File: spring/Node_t.cs
using System;

namespace spring
{
    public class Node_t
    {
        public NodeFreedom freedom;
        public NodeLoad LoadType;
        public int[] ngb;
        public float m;
        public float E;
        public float A;
        public float I;
        public float[][][] tm;
        public float[] r;
        public int NodeID;

        public Node_t(int tCounts, float[] coords, NodeFreedom _freedom, NodeLoad _LoadType, int ID, int[] neighbours, float _E, float _D)
        {
            NodeID = ID;
            freedom = _freedom;
            LoadType = _LoadType;
            r = new float[3] { 0, 0, _D / 2 };
            E = _E;
            ngb = neighbours;
            tm = new float[tCounts][][];
            for (int t = 0; t < tCounts; t++)
            {
                tm[t] = new float[Enum.GetNames(typeof(N)).Length][];
                for (int n = 0; n < tm[t].Length; n++)
                {
                    tm[t][n] = new float[3];
                }
            }
            tm[0][(int)N.p] = coords;
        }
        public void GetForces(Node_t[] Nodes, int t)
        {
            if (freedom != NodeFreedom.locked && (LoadType == NodeLoad.f|| LoadType == NodeLoad.none))
            {
                /*getting element forces*/
                foreach (var np in ngb)
                {
                    //get Fn from link between this point and np
                    float[] gFn = Element.GetFn(tm[t - 1], Nodes[np].tm[t - 1], r, A, E, I);
                    //dirty fix of dcm, just turn - to + and vs
                    vectr.Invert(ref gFn);
                    //push it to this force pull
                    vectr.Plus(ref tm[t][(int)N.f], gFn);
                }
            }
        }
        public void Integrate(int now, int before, float dt)
        {
            if (LoadType == NodeLoad.f || LoadType == NodeLoad.none)
            {
                Integr.Integrate(Integr.Types.Verlet, ref tm[now], tm[before], dt, m);
            }
            else if(LoadType == NodeLoad.p )
            {
                Integr.Integrate(Integr.Types.UVAF_P, ref tm[now], tm[before], dt, m);
            }
        }

        //private void GetLoad(int t, float[] load)
        //{
        //    switch (LoadType)
        //    {
        //        case N.p:
        //            tm[t][(int)LoadType] = load;
        //            break;

        //        case N.u:
        //            tm[t][(int)LoadType] = load;
        //            break;

        //        case N.v:
        //            break;

        //        case N.a:
        //            break;

        //        case N.b:
        //            break;

        //        case N.f:
        //            tm[t][(int)LoadType] = vectr.Plus(tm[t][(int)LoadType], load);
        //            break;

        //        default:
        //            break;
        //    }
        //}
    }
}

File: mf/Node_t.cpp
#include "Node_t.h"
#include "maf.h"
#include "dcm_t.h"
#include "integr.h"

Node_t::Node_t(int tCounts, xyz_t coords, xyz_t _radiusPoint, NodeFreedom _freedom, NodeLoad _LoadType, int _ID, std::vector<int> _Neigs)
{
    ID = _ID;
    freedom = _freedom;
    LoadType = _LoadType;
    Neigs = _Neigs;
    F.reserve(tCounts);
    deriv.reserve(tCounts);
    for (int i = 0; i < tCounts; i++)
    {
        F.push_back(xyz_t());
        deriv.push_back(deriv_t());
    }
    deriv[0].p = coords;
    deriv[0].a.y = _g;
    deriv[0].v.y = _g * 5E-06f;
    radiusPoint = _radiusPoint;
}
void Node_t::CalcAccel(int t, float m)
{
    if (LoadType == NodeLoad::none || LoadType == NodeLoad::f)
    {
        deriv[t].a.x = F[t].x / m;
        deriv[t].a.y = F[t].y / m;
        deriv[t].a.z = F[t].z / m;//has to be different
    }

}
void Node_t::GetForces(Rope_t* model, int t, float m, float c)
{
    //xyz_t Fg = new xyz_t();
    //Fg.y = m * maf._g;
    //F[t].Plus(Fg);

    xyz_t Fd;
    Fd.x = 0 - (c * deriv[t - 1].v.x);
    Fd.y = 0 - (c * deriv[t - 1].v.y);
    Fd.z = 0 - (c * deriv[t - 1].v.z);
    F[t].Plus(Fd);
    /*getting element forces*/
    for (auto neigNode : Neigs)
    {
        //getting position of link according base point
        xyz_t LinkPos;
        LinkPos.Minus(deriv[t - 1].p, (model->GetNodeRef(neigNode))->deriv[t - 1].p);
        //getting DCM for this link
        dcm_t* dcm = new dcm_t(LinkPos, radiusPoint);
        xyz_t gFn;
        //convert Fn to global coords and return
        dcm->ToGlob((model->GetElemRef(ID, neigNode))->F[t], gFn);
        //push it to this force pull
        F[t].Plus(gFn);
    }
}
void Node_t::GetPhysicParam(Rope_t* rope, int t, float Re, float& m, float& c)
{
    for (auto neigNode : Neigs)
    {
        float mElem = 0;
        float cElem = 0;
        (rope->GetElemRef(ID, neigNode))->GetPhysicParam(rope, t, Re, mElem, cElem);
        c += cElem;
        m += mElem;
    }
    //m = 1E-04f;
    c /= Neigs.size();
    if (m <= 0)
    {
        throw "Calculated mass of node can't be eaqul to zero";
    }
}
void Node_t::Integrate(int now, int before, float dt)
{
    switch (LoadType)
    {
    case NodeLoad::p:
        Integr::EulerExplBackward(deriv[now], deriv[before], dt);
        break;
    case NodeLoad::none:
        Integr::VerletForward(deriv[now], deriv[before], dt);
        break;
    case NodeLoad::f:
        Integr::VerletForward(deriv[now], deriv[before], dt);
        break;
    default:
        throw "Unexpected load type";
    }
}